src/firebase/CRUDServices/imageService.js

```diff
@@ -0,0 +1,117 @@
+  'projects/cs6510-spr2021/subscriptions/sie-image-processing-result-test-sub';
+  }
+
+    // singleton instance.
+    static sieInstance = null;
+
+    /**
+     * creating singleton instance.
+     * @return {ImageService} instance
+     */
+    static getInstance = () => {
+      if (self.sieInstance == null) {
+        self.sieInstance = new ImageService();
+      }
+      return self.sieInstance;
+    }
+
+    /**
+     * Put a self image into the cloud storage.
+     * @param {String} userName user name.
+     * @param {String} experimentId experiment id.
+     * @param {File} image self image file.
+     * @return {Boolean} succeed or not.
+     */
+    postImage = async (userName, experimentId, image) => {
+      const folderPath = userName + '-' + experimentId + '/';
+      const imagePath = folderPath + image.name;
+      const rawImageRef = self.sieRawImagesBucketRef.child(imagePath);
+      rawImageRef.put(image).then(() => {
+        return true;
+      }).catch((error) => {
+        return false;
+      });
+    }
+
+    /**
+     * Get all existing self images in the messaging queue.
+     * @param {String} userName user name.
+     * @param {String} experimentId experiment id.
+     * @return {Image[]} array of self image files.
+     */
+    getImages = async (userName, experimentId) => {
+      // Creates a client; cache this for further use
+      pubSubClient = new PubSub();
+      // References an existing subscription
+      const subscription = pubSubClient.subscription(self.subscriptionName);
+
+      // Create an event handler to handle messages
+      const messageArray = [];
+      const messageHandler = (message) => {
+        messageArray.push(message.data);
+        // "Ack" (acknowledge receipt of) the message
+        message.ack();
+      };
+
+      // Listen for new messages until timeout is hit
+      subscription.on('message', messageHandler);
+
+      setTimeout(() => {
+        subscription.removeListener('message', messageHandler);
+      }, timeout * 1000);
+
+      // loop through the messages to locate those processed images
+      // message format: [userName]-[experimentId]-[status]
+      for (let index = 0; index < messageArray.length; index++) {
+        const message = messageArray[index];
+        const [msgUserName, msgExperimentId, status] = message.split('-');
+
+        if (msgUserName == userName && msgExperimentId == experimentId) {
+          if (status == 'complete') {
+            const folderPath = userName + '-' + experimentId + '/';
+            const processedImagesRef =
+            self.sieProcessedImagesBucketRef.child(folderPath);
+            // TODO: download directly in the browser for all processed images
+            processedImagesRef.getDownloadURL().then((url) => {
+              const xhr = new XMLHttpRequest();
+              xhr.responseType = 'blob';
+              // xhr.onload = (event) => {
+              //   const blob = xhr.response;
+              // };
```

Comment:
please remove comments

Reply:
will do, got a question here though, what will we do on the firebase side when we get notified that the image processing is completed? Shall we just store the complete message in the firestore or download those processed images in cloud storage (not likely since tara said it will be sent to another cloud function to analyze if I remember correctly?)

Reply:
you're downloading the stimuli here which are what we will use for the experiment. And you should download it from `sie-stimuli/participant_id`. Sorry I haven't got a chance to update design doc

Reply:
Got it, shall we download it in the browser directly and generate a zip file of all images? If so we may need to add filesaver.js to do that.

Reply:
I'm not sure here. As long as the experiment has images for participants to select and it's reasonably fast, we're good